--- ./src/racoon/Makefile.am.link	2010-04-28 05:56:52.000000000 +0000
+++ ./src/racoon/Makefile.am	2010-04-28 06:04:37.000000000 +0000
@@ -49,6 +49,7 @@
 racoonctl_LDADD = libracoon.la ../libipsec/libipsec.la 
 
 libracoon_la_SOURCES = kmpstat.c vmbuf.c sockmisc.c misc.c
+libracoon_la_LIBADD = $(top_builddir)/src/libipsec/libipsec.la
 
 plainrsa_gen_SOURCES = plainrsa-gen.c plog.c \
 	crypto_openssl.c logger.c 
--- ./src/racoon/kmpstat.c.link	2010-04-28 06:11:19.000000000 +0000
+++ ./src/racoon/kmpstat.c	2010-04-28 06:12:03.000000000 +0000
@@ -89,6 +89,7 @@
 #include "admin_var.h"
 #include "ipsec_doi.h"
 
+char *adminsock_path = ADMINSOCK_PATH;
 u_int32_t racoonctl_interface = RACOONCTL_INTERFACE;
 u_int32_t racoonctl_interface_major = RACOONCTL_INTERFACE_MAJOR;
 
